src/mon/OSDMonitor.cc

```diff
@@ -8417,6 +8458,204 @@ bool OSDMonitor::prepare_command_impl(MonOpRequestRef op,
+        osd_first = parse_osd_id(first.c_str(), &ss);
+        osd_last = parse_osd_id(last.c_str(), &ss);
+      } else {
+        // try traditional single osd way
+        osd_first = parse_osd_id(idvec[j].c_str(), &ss);
+        osd_last = osd_first;
+      }
+
+      if (osd_first < 0 || osd_last < 0 || osd_first > osd_last) {
+        // ss has reason for failure
+        ss << ", unable to parse osd id[range]:\"" << idvec[j] << "\". ";
+        err = -EINVAL;
+        continue;
+      }
+
+      for (auto osd = osd_first; osd <= osd_last; osd++) {
+
+        if (!osdmap.exists(osd)) {
+          ss << "osd." << osd << " does not exist. ";
+          continue;
+        }
+
+        if (noout) {
+          if (osdmap.is_out(osd)) {
+            ss << "osd." << osd << " is already out. ";
+            continue;
+          }
+
+          if (osdmap.is_noout(osd)) {
+            ss << "osd." << osd << " is already marked noout. ";
+            continue;
+          }
+
+          pending_inc.new_state[osd] |= CEPH_OSD_NOOUT;
+          any = true;
+        } else {
+          // nodown
+          if (osdmap.is_down(osd)) {
+            ss << "osd." << osd << " is already down. ";
+            continue;
+          }
+
+          if (osdmap.is_nodown(osd)) {
+            ss << "osd." << osd << " is already marked nodown. ";
+            continue;
+          }
+
+          pending_inc.new_state[osd] |= CEPH_OSD_NODOWN;
+          any = true;
+        }
+      }
+    }
+
+    if (any) {
+      getline(ss, rs);
+      wait_for_finished_proposal(op, new Monitor::C_Command(mon, op, err, rs,
+                                 get_last_committed() + 1));
+      return true;
+    }
+  } else if (prefix == "osd rm-noout" ||
+             prefix == "osd rm-nodown") {
+
+    bool any = false;
+    bool stop = false;
+    bool noout = prefix == "osd rm-noout";
+
+    vector<string> idvec;
+    cmd_getval(g_ceph_context, cmdmap, "ids", idvec);
+
+    for (unsigned j = 0; j < idvec.size() && !stop; j++) {
+      vector<int> osds;
+
+      // for removal, wildcald "any" is also supported,
+      // e.g., to quickly cancel all previous noout[nodown] osds
+      if (j == 0 && idvec[0] == "any") {
+        if (noout) {
+          osdmap.get_noout_osds(&osds);
+        } else {
+          osdmap.get_nodown_osds(&osds);
+        }
+
+        // cancel pending noout[nodown] requests too,
+        // if there is any
+        for (auto &p : pending_inc.new_state) {
+          if (noout) {
+            if (!osdmap.is_noout(p.first) &&
+                (p.second & CEPH_OSD_NOOUT) != 0) {
+              p.second &= ~CEPH_OSD_NOOUT;
+              any = true;
+            }
+          } else {
+            if (!osdmap.is_nodown(p.first) &&
+                (p.second & CEPH_OSD_NODOWN) != 0) {
+              p.second &= ~CEPH_OSD_NODOWN;
+              any = true;
+            }
+          }
+        }
```

Comment:
I wonder if we'd be better served with new helper functions like _pending_osd_state_{set,clear}(int osd, unsigned bit) that hide this ugliness.  I see that 'osd down' and most of the other places that touch new_state do bother to check pending_inc and can clobber a racing update.  That'll simplify this quite a bit.  (Can leave fixing the other would-be users as a follow-up)